src/base/http/requestparser.cpp

```diff
@@ -218,15 +180,10 @@ bool RequestParser::parseRequestLine(const QString &line)
     m_request.method = match.captured(1);
 
     // Request Target
-    const QUrl url = QUrl::fromEncoded(match.captured(2).toLatin1());
-    m_request.path = url.path();
-
-    // parse queries
-    QListIterator<QStringPair> i(QUrlQuery(url).queryItems());
-    while (i.hasNext()) {
-        const QStringPair pair = i.next();
-        m_request.gets[pair.first] = pair.second;
-    }
+    const QByteArray decodedUrl {QByteArray::fromPercentEncoding(match.captured(2).toLatin1())};
+    const int sepPos = decodedUrl.indexOf('?');
+    m_request.path = (sepPos == -1 ? decodedUrl : midView(decodedUrl, 0, sepPos));
```

Comment:
how about midView -> QString::fromLatin1 ?

Reply:
Why `fromLatin1()`? Path can contain Utf8 characters.

Reply:
Changed with:
```c++
m_request.path = QString::fromUtf8(decodedUrl.constData(), (sepPos == -1 ? decodedUrl.size() : sepPos));
```

Reply:
>Path can contain Utf8 characters.

Partially right, AFAIK, the utf8 characters can be encoded in Punycode or percent-encoding, in both cases the representation is still latin-1 chars only.

https://stackoverflow.com/questions/22357509/can-urls-have-utf-8-characters

Besides, `decodedUrl` is constructed with `...toLatin1()`, so no utf8 chars will occur here.

Reply:
@Chocobo1, please look at the algorithm carefully!

URL can contains only Latin-1 characters. Any other characters should be encoded.
>Besides, decodedUrl is constructed with ...toLatin1(), so no utf8 chars will occur here.

We obtain URL from request line using toLatin1() and then use `QByteArray::fromPercentEncoding()` so decodedUrl can contain any kind of data in it.
Then we split it into Path and Query.
Path is expected to be Utf8 string so we decode it as Utf8.
Query can contain any binary data (depending on web application) so we store it as raw byte array for now. At application layer it will be decoded in appropriate way.

Reply:
You are right, pardon me, I realize that 2 mins later after my reply.